### PR TITLE
SALTO-1048 Limit concurrent salesforce API calls

### PR DESCRIPTION
**Pull request by netama**
Initial support for throttling API calls in the client - currently supports `list`, `retrieve`, `read` (each in its own bucket) and `total` (for requests that passed the initial per-request throttling), but can be extended easily to additional requests.
Each of these can be disabled by setting to `-1` in the configuration (by default the behavior is the same as today - only `retrieve` has a rate limit).
Also adjusted the logs so we can see when a message was `enqueued` separately from when it actually `started`.

---
_Release Notes_
Salesforce - Extend configuration options for limiting concurrent API calls, in order to address `REQUEST_LIMIT_EXCEEDED` errors on large workspaces.

_Upgrade Instructions_
The `maxConcurrentRetrieveRequests` configuration parameter is no longer used. If it was modified from the original value, add the following section under `salesforce.nacl` (with the value from your configuration):
```
  client = {
    maxConcurrentApiRequests = {
      retrieve = <value>
    }
  }
```
If `maxConcurrentApiRequests` is not set, the max concurrent `retrieve` requests will default to 3.

### CODE REVIEW

File: packages/cli/e2e_test/NACL/salto.config/salesforce.nacl [hunk 1]
```diff
@@ -10,7 +10,9 @@ salesforce {
   instancesRegexSkippedList = [
     "^EmailTemplate.MarketoEmailTemplates",
   ]
-  maxConcurrentRetrieveRequests = 3
```
Comment (netama) [inline packages/cli/e2e_test/NACL/salto.config/salesforce.nacl:13]: I added upgrade instructions, not sure how frequently this is customized - is it worth adding a temporary fallback to look at the old config?

Comment (omrilit) [reply packages/cli/e2e_test/NACL/salto.config/salesforce.nacl:13]: I don't think so. I think that the upgrade instructions should also state what is the actual behavior if the user doesn't care about this value and never edited it.

File: packages/salesforce-adapter/src/client/client.ts [hunk 2]
```diff
@@ -241,6 +245,22 @@ const createConnectionFromCredentials = (
   return realConnection(creds.isSandbox, options)
 }
 
+const createRateLimitersFromConfig = (
+  rateLimit: RateLimitConfig,
+): Record<RateLimitBucketName, Bottleneck> => {
+  const toLimit = (
+    num: number | undefined
+  ): number | undefined => (num && num < 0 ? undefined : num)
```
Comment (omrilit) [inline packages/salesforce-adapter/src/client/client.ts:253]: What is the expected behavior if the user sets a certain call to have limit=0?
Maybe we should handle it manually? (set to 1 / -1 / throw exception?)

File: packages/salesforce-adapter/e2e_test/custom_object_instances.test.ts [hunk 3]
```diff
@@ -52,8 +52,10 @@ describe('custom object instances e2e', () => {
   }
   beforeAll(async () => {
     credLease = await testHelpers().credentials()
-    const adapterParams = realAdapter({ credentials:
-      new UsernamePasswordCredentials(credLease.value) }, filtersContext)
+    const adapterParams = realAdapter({
+      credentials: new UsernamePasswordCredentials(credLease.value),
+      rateLimit: { total: -1, retrieve: 3, read: -1, list: -1 },
```
Comment (omrilit) [inline packages/salesforce-adapter/e2e_test/custom_object_instances.test.ts:57]: nit: use RATE_LIMIT_UNLIMITED_MAX_CONCURRENT_REQUESTS

File: packages/salesforce-adapter/src/types.ts [hunk 4]
```diff
@@ -221,6 +228,36 @@ const dataManagementType = new ObjectType({
...
+    total: {
+      type: BuiltinTypes.NUMBER,
+      annotations: {
+        [CORE_ANNOTATIONS.RESTRICTION]: createRestriction({ min: -1, max: 100 }),
+      },
+    },
+    retrieve: {
+      type: BuiltinTypes.NUMBER,
+      annotations: {
+        [CORE_ANNOTATIONS.RESTRICTION]: createRestriction({ min: -1, max: 100 }),
```
Comment (omrilit) [inline packages/salesforce-adapter/src/types.ts:243]: Seems like there was a restriction to max=25 before (although we don't enforce it), did you intentionally change it?

File: packages/salesforce-adapter/test/adapter.creator.test.ts [hunk 5]
```diff
@@ -119,6 +119,12 @@ describe('SalesforceAdapter creator', () => {
           isSandbox: false,
           apiToken: 'myToken',
         }),
+        rateLimit: {
+          list: -1,
```
Comment (omrilit) [inline packages/salesforce-adapter/test/adapter.creator.test.ts:123]: nit: use RATE_LIMIT_UNLIMITED_MAX_CONCURRENT_REQUESTS

File: packages/salesforce-adapter/test/client.test.ts [hunk 6]
```diff
@@ -60,16 +60,24 @@ describe('salesforce client', () => {
...
+      password: '',
+      isSandbox: true,
+    }),
+    retryOptions: {
+      maxAttempts: 4, // try 4 times
+      retryDelay: 100, // wait for 100ms before trying again
+      retryStrategy: RetryStrategies.NetworkError, // retry on network errors
+    },
+    rateLimit: {
+      total: -1, // unlimited total requests
```
Comment (omrilit) [inline packages/salesforce-adapter/test/client.test.ts:75]: nit: use RATE_LIMIT_UNLIMITED_MAX_CONCURRENT_REQUESTS

File: packages/salesforce-adapter/test/client.ts [hunk 7]
```diff
@@ -26,6 +26,12 @@ const mockClient = (): { connection: MockInterface<Connection>; client: Salesfor
       password: 'mockPassword',
       isSandbox: false,
     },
+    rateLimit: {
+      total: -1,
```
Comment (omrilit) [inline packages/salesforce-adapter/test/client.ts:30]: nit: use RATE_LIMIT_UNLIMITED_MAX_CONCURRENT_REQUESTS

**Review by omrilit — APPROVED** [review]
Looks great!

File: packages/salesforce-adapter/src/types.ts (hunk 4 above)
Comment (netama) [reply packages/salesforce-adapter/src/types.ts:243]: yes (though I did find out when testing that it doesn't matter, like you said :) ) - my assumption was that the old value was set with `retrieve` in mind (which has large requests). For `list` / `read` most requests are very quick and don't count towards the limit, so in terms of restricting the possible values I thought it was better to leave some more room for "not unlimited but not too restricted". Not sure if 100 is high enough for that

File: packages/salesforce-adapter/src/client/client.ts (hunk 2 above)
Comment (netama) [reply packages/salesforce-adapter/src/client/client.ts:253]: Good point - because we went with `-1` as unlimited I thought to just leave `0` as "always block", but you're right this might end up being confusing. I'm leaning towards an exception but will check with others, any preference?

Comment (omrilit) [reply packages/salesforce-adapter/src/client/client.ts:253]: This is really an edge case so I think an exception is fine

File: packages/salesforce-adapter/test/client.test.ts (hunk 6 above)
Comment (netama) [reply packages/salesforce-adapter/test/client.test.ts:75]: I'm removing the others that are the same as the default, but I prefer to keep this one explicit (wanted to have the two variants explicitly in one of the tests)

File: packages/salesforce-adapter/src/client/client.ts (hunk 2 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/client/client.ts:253]: we could avoid this edge case by making 0 the value for unlimited...

File: packages/salesforce-adapter/e2e_test/custom_object_instances.test.ts (hunk 3 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/e2e_test/custom_object_instances.test.ts:57]: IMO we do not have to use the constants where we have types.

File: packages/salesforce-adapter/src/types.ts (hunk 4 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/types.ts:243]: do we need a `max` at all?

File: packages/salesforce-adapter/src/client/client.ts (hunk 2 above)
Comment (netama) [reply packages/salesforce-adapter/src/client/client.ts:253]: seems that's not what we want though (see the last comment on the ticket)

File: packages/salesforce-adapter/src/types.ts (hunk 4 above)
Comment (netama) [reply packages/salesforce-adapter/src/types.ts:243]: wasn't sure. since this is not enforced anyway I'll just remove it

**Comment by netama** [comment]
moved inside the client config based on a request from @ori-moisis
